code.js

```diff
@@ -1,3 +1,46 @@
 function tsp_ls(distance_matrix) {
-    return -1;
+  let n = distance_matrix.length;
+  let memo = new Set();
+  let minRoute = Array.from({ length: n }, (v, i) => i).sort((a, b) => 0.5 - Math.random());
+  let minCost = calcCost(minRoute, distance_matrix);
+  let improving = true;
+  
+  while(improving) {
+    let routePath = [...minRoute]; // Get a deep copy of min route
+    let r = Math.floor(Math.random() * (n - 1)) + 1; // Random Number in the range 1 to n - 1
+    let l = Math.floor(Math.random() * (n - r - 1)) + (r + 1); // Random number in the range r + 1 to n
+    
+    // If the swap was already done don`t undo it
+    if (memo.has(`${r}, ${l}`)) continue;
```

Comment:
But what about previous swaps?

Reply:
I didn't think about that but I think the way to correct it would be to just use the entire path as a key.
    let routeKey = routePath.join(",");
    if (memo.has(routeKey)) continue;
    memo.add(routeKey);